### ISSUE TITLE
Cannot allocate memory

### ISSUE THREAD

**Issue by ethanmdavidson** [hashicorp/vagrant#11420]
### Vagrant version
Vagrant 2.2.7

### Host operating system
Windows 10 (1809)
Virtualbox 6.1.4

### Guest operating system
Oracle 7.7

### Vagrantfile
```ruby
# -*- mode: ruby -*-
# vi: set ft=ruby :

Vagrant.configure("2") do |config|
  config.vm.box = "bento/oracle-7"
  config.vm.define "vagrant-test" do | devapp|
	  # change memory size
	  devapp.vm.provider "virtualbox" do |v|
		v.memory = 4096
		v.cpus = 4
	  end
    end
end
```

### Debug output
https://gist.github.com/ethanmdavidson/ba1234eca31995be70a07676ec2a584e

### Expected behavior
The public key should be successfully inserted and the vm should finish upping.

### Actual behavior
An error is encountered (```C:/HashiCorp/Vagrant/embedded/gems/2.2.6/gems/net-scp-1.2.1/lib/net/scp.rb:398:in `await_response_state': scp: /tmp/vagrant-insert-pubkey-1572381127: Cannot allocate memory (RuntimeError)```) while inserting the public key and the vm is deleted.

### Steps to reproduce
1. using the included vagrantfile, run `vagrant up`

Note that it doesn't seem to happen every time. It was happening every time I upped the included vagrantfile, but when I tried to run in debug mode it worked the first time, but then encountered the error the next time I ran it. My host machine has plenty of ram (32GB total, only using ~13GB) and the error seems to occur more often the higher the memory is set (e.g. happens a lot with `v.memory = 8192`, doesn't seem to ever happen with `v.memory=1024`).

Also important to note is that this error doesn't seem to occur when using a different oracle 7 box, e.g. `generic/oracle7`. It was only after we switched to `bento/oracle-7` that we started seeing this error. I'm opening this issue here because it seems to be falling on vagrant's side, but I can open an issue with bento if it's determined to be an issue with the way they're creating boxes.

### References

Seems to be the same error as #11166 (and the guest OS I'm running, Oracle 7, is based on the guest OS they were running, CentOS 7).

**Comment by ethanmdavidson** [comment]
Also seems much more likely to happen if you already have another vagrant box created? Sorry, still pinning down the exact circumstances required to reliably trigger this.

**Comment by briancain** [comment]
Hey there @ethanmdavidson - First, I recommend not setting your guest to use 4 CPUs unless your host machine has a lot more than that. Often giving a guest all of the available CPUs will decrease performance, rather than increase it (intuitive, I know :D) Second, when you receive the error, are you able to check the guest to see how much available memory there is? Also I will say that if this seems to only happen with the bento box, I recommend opening an issue on their side. It's possible that it's just an issue with that specific box.

**Comment by ethanmdavidson** [comment]
Unfortunately this issue happens when vagrant tries to insert the ssh key, which happens pretty much as soon as the machine boots for the first time. I've tried manually sshing in with the default password but I'm not quick enough to get in before vagrant encounters the error and starts tearing things down. 

Do you have any recommendations for how to monitor the memory before the box is up?

I've also noticed that adding `config.ssh.insert_key=false` to the vagrant file seems to allow everything to complete successfully. Definitely seems to be some problem in the insert key step.

**Comment by briancain** [comment]
@ethanmdavidson - When you disable the ssh key insert, are you able to go onto the guest and see how much available RAM there is after Vagrant finishes bringing up the guest? Does it seem to report the amount Vagrant allocated for it? I unfortunately haven't been able to reproduce this, even with multiple guests running and various amounts of RAM and CPU for the oracle guest :disappointed: If it's not too much trouble, do you get the same behavior on a previous version of VirtualBox...say version 6.0 instead of 6.1?

**Comment by ethanmdavidson** [comment]
Don't worry too much about it if you can't reproduce it; I can hardly reproduce it myself. If I run vagrant up several times in a row, it seems to only fail half the time.

Even when it fails, the virtualbox manager reports the correct memory setting. When it succeeds (whether randomly or because insert_key is disabled) it also reports the correct amount of memory. 
```
[vagrant@localhost ~]$ free -h
              total        used        free      shared  buff/cache   available
Mem:           7.5G        119M        7.2G         16M        219M        7.2G
Swap:          2.0G          0B        2.0G
```

I installed virtualbox 6.0.18 and experienced the same symptoms.

Without being able to reproduce it, I know this will be very difficult to solve. I can't even say for sure that it's an issue with vagrant, and it's also not a huge issue (whenever it fails I can just up it again and it'll usually work the second time) so I won't hold it against you if you don't spend a ton of time on this.

**Comment by makker-nl** [comment]
I've got the same error but a bit later down the flow, when mounting synched folders. In some cases also later when running provisioning:
```
d:\Projects\vagrant\ol77_tf>vagrant up
Bringing machine 'darwin-tf' up with 'virtualbox' provider...
==> darwin-tf: Importing base box 'ol77GUIv1.0'...
==> darwin-tf: Matching MAC address for NAT networking...
==> darwin-tf: Setting the name of the VM: OL7U7_TF
==> darwin-tf: Clearing any previously set network interfaces...
==> darwin-tf: Preparing network interfaces based on configuration...
    darwin-tf: Adapter 1: nat
==> darwin-tf: Forwarding ports...
    darwin-tf: 22 (guest) => 2222 (host) (adapter 1)
==> darwin-tf: Running 'pre-boot' VM customizations...
==> darwin-tf: Booting VM...
==> darwin-tf: Waiting for machine to boot. This may take a few minutes...
    darwin-tf: SSH address: 127.0.0.1:2222
    darwin-tf: SSH username: vagrant
    darwin-tf: SSH auth method: private key
    darwin-tf:
    darwin-tf: Vagrant insecure key detected. Vagrant will automatically replace
    darwin-tf: this with a newly generated keypair for better security.
    darwin-tf:
    darwin-tf: Inserting generated public key within guest...
    darwin-tf: Removing insecure key from the guest if it's present...
    darwin-tf: Key inserted! Disconnecting and reconnecting using new SSH key...
==> darwin-tf: Machine booted and ready!
[darwin-tf] GuestAdditions 6.1.2 running --- OK.
==> darwin-tf: Checking for guest additions in VM...
==> darwin-tf: Mounting shared folders...
    darwin-tf: /vagrant => D:/Projects/vagrant/ol77_tf
    darwin-tf: /media/sf_Stage => D:/Projects/vagrant/Stage
The following SSH command responded with a non-zero exit status.
Vagrant assumes that this means the command failed!

mkdir -p /media/sf_Stage

Stdout from the command:



Stderr from the command:

mkdir: cannot create directory ‘/media/sf_Stage’: Cannot allocate memory

d:\Projects\vagrant\ol77_tf>

```

**Comment by chrisroberts** [comment]
There have been past issues with vbox where errors can crop up when attempting to allocate more memory from the host and the host's memory is fragmented. One solution may be to pre-allocate the memory when starting the guest:

```ruby
Vagrant.configure("2") do |config|
  ...
  config.vm.provider :virtualbox do |vb|
    vb.customize ["setextradata", :id, "VBoxInternal/RamPreAlloc", 1]
  end
end
```

Let me know if that helps at all resolving this issue.

**Comment by awltux** [comment]
I'm seeing this issue after upgrading 
- Vagrant 2.2.6 to 2.2.7
- VirtualBox 6.0.16 to 6.1.6 r137129
Intermittent. Always happens during 'vagrant up' and it seems to be after Virtualbox mounts.

**Comment by makker-nl** [comment]
I tried the suggestion of ChrisRoberts, but then it failed in even creating the box.
Upgraded to Vagrant 229 en VirtualBox 6.1.6 but get the same error:
```
error: db5 error(12) from dbenv->open: Cannot allocate memory
error: cannot open Packages index using db5 - Cannot allocate memory (12)
```

**Comment by makker-nl** [comment]
I made sure that vagrant can logon using a password in ssh and provided username/password:
  config.ssh.username="vagrant"
  config.ssh.password="vagrant"

Now it seems to work.  It seems to me that using a ssh-key somehow drives the memory allocation problems. Can't see how that makes sense, but somehow that seems to be the case.

**Comment by briancain** [comment]
> Don't worry too much about it if you can't reproduce it; I can hardly reproduce it myself. If I run vagrant up several times in a row, it seems to only fail half the time.
> 
> Even when it fails, the virtualbox manager reports the correct memory setting. When it succeeds (whether randomly or because insert_key is disabled) it also reports the correct amount of memory.
> 
> ```
> [vagrant@localhost ~]$ free -h
>               total        used        free      shared  buff/cache   available
> Mem:           7.5G        119M        7.2G         16M        219M        7.2G
> Swap:          2.0G          0B        2.0G
> ```
> 
> I installed virtualbox 6.0.18 and experienced the same symptoms.
> 
> Without being able to reproduce it, I know this will be very difficult to solve. I can't even say for sure that it's an issue with vagrant, and it's also not a huge issue (whenever it fails I can just up it again and it'll usually work the second time) so I won't hold it against you if you don't spend a ton of time on this.

Thanks @ethanmdavidson . I can't reproduce it either, so I'm going to go ahead and close this.

Since others in this thread seem to be experiencing a different memory allocation issue separate from the original issue, I recommend opening a new issue. I'm going to go ahead and close this one since I cannot reproduce the original posters problem. Thanks!
